ClassCastException in Lincheck

Lincheck version 2.34

Trace: 

```
Wow! You've caught a bug in Lincheck.
We kindly ask to provide an issue here: https://github.com/JetBrains/lincheck/issues,
attaching a stack trace printed below and the code that causes the error.

Exception stacktrace:
java.lang.ClassCastException

org.jetbrains.kotlinx.lincheck.LincheckAssertionError: 
Wow! You've caught a bug in Lincheck.
We kindly ask to provide an issue here: https://github.com/JetBrains/lincheck/issues,
attaching a stack trace printed below and the code that causes the error.

Exception stacktrace:
java.lang.ClassCastException

	at org.jetbrains.kotlinx.lincheck.LinChecker$check$1.invoke(LinChecker.kt:48)
	at org.jetbrains.kotlinx.lincheck.LinChecker$check$1.invoke(LinChecker.kt:47)
	at org.jetbrains.kotlinx.lincheck.LinChecker.checkImpl$lincheck(LinChecker.kt:67)
	at org.jetbrains.kotlinx.lincheck.LinChecker.check(LinChecker.kt:47)
	at org.jetbrains.kotlinx.lincheck.LinChecker$Companion.check(LinChecker.kt:195)
	at org.jetbrains.kotlinx.lincheck.LinCheckerKt.check(LinChecker.kt:295)
	at reproClassCast.LinCheckTest.stressTest(Unknown Source)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.runTestClass(JUnitTestClassExecutor.java:112)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:58)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:40)
	at org.gradle.api.internal.tasks.testing.junit.AbstractJUnitTestClassProcessor.processTestClass(AbstractJUnitTestClassProcessor.java:60)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:52)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at jdk.proxy1/jdk.proxy1.$Proxy2.processTestClass(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker$2.run(TestWorker.java:176)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.executeAndMaintainThreadName(TestWorker.java:129)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:100)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:60)
	at org.gradle.process.internal.worker.child.ActionExecutionWorker.execute(ActionExecutionWorker.java:56)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:119)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:66)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)
```

DS under test:
```Kotlin
package reproClassCast

import java.util.concurrent.atomic.*
import kotlin.math.absoluteValue

class ConcurrentHashTable<K : Any, V : Any>(initialCapacity: Int) {
    private val table = AtomicReference(Table(initialCapacity))

    fun put(key: K, value: V): V? {
        return table.get().put(key, value)
    }

    fun get(key: K): V? {
        return table.get().get(key)
    }

    fun remove(key: K): V? {
        return table.get().remove(key)
    }

    data class Fixed(val value: Any)

    inner class Table(val capacity: Int) {

        var resizeDone = false
        val keys = AtomicReferenceArray<Any?>(capacity)
        val values = AtomicReferenceArray<Any?>(capacity)
        val nextTable = AtomicReference<Table?>(null)

        fun put(key: K, value: V): V? {
            var index = index(key)
            repeat(MAX_PROBES) {
                while (true) {
                    val curKey = keys[index]
                    when (curKey) {
                        null -> {
                            if (!keys.compareAndSet(index, null, key)) continue
                            if (!values.compareAndSet(index, null, value)) continue
                            return null
                        }
                        key -> {
                            while (true) {
                                val currentValue = values[index]
                                if (currentValue == MOVED) {
                                    return resize { tab ->
                                        tab.put(key, value)
                                    } as V?
                                }
                                if (currentValue is Fixed) {
                                    return resize { tab ->
                                        tab.put(key, value)
                                    } as V?
                                }
                                if (!values.compareAndSet(index, currentValue, value)) continue
                                return currentValue as V?
                            }
                        }
                    }
                    index = (index + 1) % capacity
                    break
                }
            }
            return resize { tab ->
                tab.put(key, value)
            } as V?
        }

        fun get(key: K): V? {
            var index = index(key)
            repeat(MAX_PROBES) {
                val curKey = keys[index]
                when (curKey) {
                    key -> {
                        val curValue = values[index]
                        if (curValue == MOVED) {
                            resize()
                            return table.get()!!.get(key)
                        }
                        if (curValue is Fixed) {
                            return curValue.value as V?
                        }
                        return values[index] as V?
                    }
                    null -> {
                        return null
                    }
                }
                index = (index + 1) % capacity
            }
            return null
        }

        fun remove(key: K): V? {
            var index = index(key)
            repeat(MAX_PROBES) {
                val curKey = keys[index]
                when (curKey) {
                    key -> {
                        while (true) {
                            val curValue = values[index]
                            if (curValue == MOVED) {
                                return resize { tab ->
                                    tab.remove(key)
                                } as V?
                            }
                            if (curValue is Fixed) {
                                return resize { tab ->
                                    tab.remove(key)
                                } as V?
                            }
                            if (!values.compareAndSet(index, curValue, null)) continue
                            return curValue as V?
                        }
                    }
                    null -> {
                        return null
                    }
                }
                index = (index + 1) % capacity
            }
            return null
        }

        fun resize(doAfter: (Table) -> Any? = { null }): Any? {
            nextTable.compareAndSet(null, Table(capacity * 2))
            val next = nextTable.get()!!

            for (i in 0 ..< capacity) {
                while (true) {
                    val curKey = keys[i]
                    val curValue = values[i]

                    // if already moved continue
                    if (curValue == MOVED) break

                    // if empty or removed -> MOVED
                    if (curValue == null) {
                        if (!values.compareAndSet(i, null, MOVED)) continue
                        break
                    }

                    // if fixed -> put value and set MOVED
                    if (curValue is Fixed) {
                        // Set value in new table
                        next.put(curKey as K, curValue.value as V)

                        // Set value to moved in this table
                        if (!values.compareAndSet(i, curValue, MOVED)) continue
                        break
                    }

                    // if normal value -> fix
                    values.compareAndSet(i, curValue, Fixed(curValue))
                }
            }
            val resultAfter = doAfter(next)
            resizeDone = true

            //Update table reference
            while (true) {
                val currentTable = table.get()
                if (currentTable.capacity < this.capacity) break
                if (currentTable.nextTable.get() == null) break
                if (currentTable.nextTable.get() != null && !currentTable.resizeDone) {
                    currentTable.resize()
                    continue
                }
                table.compareAndSet(currentTable, currentTable.nextTable.get())
            }

            return resultAfter

        }

        private fun index(key: Any) = ((key.hashCode() * MAGIC) % capacity).absoluteValue
    }
}

private const val MAGIC = -0x61c88647 // golden ratio 
private const val MAX_PROBES = 2
private val MOVED = Any()
```

Test class:

```Kotlin
package reproClassCast

import org.jetbrains.kotlinx.lincheck.annotations.Operation
import org.jetbrains.kotlinx.lincheck.check
import org.jetbrains.kotlinx.lincheck.strategy.stress.StressOptions
import kotlin.test.Test

class LinCheckTest {

    @Test
    fun stressTest() = StressOptions()
        .iterations(300)
        .invocationsPerIteration(25_000)
        .actorsBefore(1)
        .threads(3)
        .actorsPerThread(3)
        .actorsAfter(0)
        .check(this::class.java)

    
    private val hashTable = ConcurrentHashTable<Int, Int>(initialCapacity = 2)

    @Operation
    fun put(key: Int, value: Int): Int? = hashTable.put(key, value)

    @Operation
    fun get(key: Int): Int? = hashTable.get(key)

    @Operation
    fun remove(key: Int): Int? = hashTable.remove(key)

}

````

--- a/src/jvm/main/org/jetbrains/kotlinx/lincheck/Reporter.kt
+++ b/src/jvm/main/org/jetbrains/kotlinx/lincheck/Reporter.kt
@@ -615,8 +615,14 @@ private fun Throwable.isInternalLincheckBug(): Boolean {
     // we expect every Lincheck test thread to start from the Lincheck runner routines,
     // so we filter out stack trace elements of these runner routines
     val testStackTrace = stackTrace.takeWhile { LINCHECK_RUNNER_PACKAGE_NAME !in it.className }
-    // if the stack trace contains Lincheck functions, we classify it as a Lincheck bug
-    return testStackTrace.any { LINCHECK_PACKAGE_NAME in it.className }
+    // collect Lincheck functions from the stack trace
+    val lincheckStackFrames = testStackTrace.filter { LINCHECK_PACKAGE_NAME in it.className }
+    // special handling of `cancelByLincheck` primitive
+    if (lincheckStackFrames.size == 1 && "cancelByLincheck" in lincheckStackFrames[0].toString()) {
+        return false
+    }
+    // otherwise, if the stack trace contains any Lincheck functions, we classify it as a Lincheck bug
+    return lincheckStackFrames.isNotEmpty()
 }
 
 private fun StringBuilder.appendUnexpectedExceptionFailure(

--- a/src/jvm/main/org/jetbrains/kotlinx/lincheck/Reporter.kt
+++ b/src/jvm/main/org/jetbrains/kotlinx/lincheck/Reporter.kt
@@ -601,7 +601,7 @@ internal fun collectExceptionStackTraces(executionResult: ExecutionResult): Exce
         .filterIsInstance<ExceptionResult>()
         .forEachIndexed { index, exceptionResult ->
             val exception = exceptionResult.throwable
-            val isInternalLincheckBug = exception.stackTrace?.lastOrNull()
+            val isInternalLincheckBug = exception.stackTrace?.firstOrNull()
                 ?.let { LINCHECK_PACKAGE_NAME in it.className }
                 ?: false
             if (isInternalLincheckBug) {

--- a/src/jvm/main/org/jetbrains/kotlinx/lincheck/Reporter.kt
+++ b/src/jvm/main/org/jetbrains/kotlinx/lincheck/Reporter.kt
@@ -597,22 +597,20 @@ internal data class ExceptionStackTracesResult(val exceptionStackTraces: Map<Thr
  */
 internal fun collectExceptionStackTraces(executionResult: ExecutionResult): ExceptionsProcessingResult {
     val exceptionStackTraces = mutableMapOf<Throwable, ExceptionNumberAndStacktrace>()
-
-    (executionResult.initResults.asSequence()
-            + executionResult.parallelResults.asSequence().flatten()
-            + executionResult.postResults.asSequence())
+    executionResult.allResults
         .filterIsInstance<ExceptionResult>()
         .forEachIndexed { index, exceptionResult ->
             val exception = exceptionResult.throwable
-
-            val filteredStacktrace = exception.stackTrace.takeWhile { LINCHECK_PACKAGE_NAME !in it.className }
-            if (filteredStacktrace.isEmpty()) { // Exception in Lincheck itself
+            val isInternalLincheckBug = exception.stackTrace?.lastOrNull()
+                ?.let { LINCHECK_PACKAGE_NAME in it.className }
+                ?: false
+            if (isInternalLincheckBug) {
                 return InternalLincheckBugResult(exception)
             }
-
-            exceptionStackTraces[exception] = ExceptionNumberAndStacktrace(index + 1, filteredStacktrace)
+            val stackTrace = exception.stackTrace
+                .filter { LINCHECK_PACKAGE_NAME !in it.className }
+            exceptionStackTraces[exception] = ExceptionNumberAndStacktrace(index + 1, stackTrace)
         }
-
     return ExceptionStackTracesResult(exceptionStackTraces)
 }
 

--- a/src/jvm/main/org/jetbrains/kotlinx/lincheck/Reporter.kt
+++ b/src/jvm/main/org/jetbrains/kotlinx/lincheck/Reporter.kt
@@ -601,10 +601,7 @@ internal fun collectExceptionStackTraces(executionResult: ExecutionResult): Exce
         .filterIsInstance<ExceptionResult>()
         .forEachIndexed { index, exceptionResult ->
             val exception = exceptionResult.throwable
-            val isInternalLincheckBug = exception.stackTrace?.firstOrNull()
-                ?.let { LINCHECK_PACKAGE_NAME in it.className }
-                ?: false
-            if (isInternalLincheckBug) {
+            if (exception.isInternalLincheckBug()) {
                 return InternalLincheckBugResult(exception)
             }
             val stackTrace = exception.stackTrace
@@ -612,6 +609,14 @@ internal fun collectExceptionStackTraces(executionResult: ExecutionResult): Exce
             exceptionStackTraces[exception] = ExceptionNumberAndStacktrace(index + 1, stackTrace)
         }
     return ExceptionStackTracesResult(exceptionStackTraces)
+}
+
+private fun Throwable.isInternalLincheckBug(): Boolean {
+    // we expect every Lincheck test thread to start from the Lincheck runner routines,
+    // so we filter out stack trace elements of these runner routines
+    val testStackTrace = stackTrace.takeWhile { LINCHECK_RUNNER_PACKAGE_NAME !in it.className }
+    // if the stack trace contains Lincheck functions, we classify it as a Lincheck bug
+    return testStackTrace.any { LINCHECK_PACKAGE_NAME in it.className }
 }
 
 private fun StringBuilder.appendUnexpectedExceptionFailure(

--- a/src/jvm/main/org/jetbrains/kotlinx/lincheck/Utils.kt
+++ b/src/jvm/main/org/jetbrains/kotlinx/lincheck/Utils.kt
@@ -283,3 +283,4 @@ internal inline fun <R> ParallelThreadsRunner.runOutsideIgnoredSection(currentTh
 }
 
 internal const val LINCHECK_PACKAGE_NAME = "org.jetbrains.kotlinx.lincheck."
+internal const val LINCHECK_RUNNER_PACKAGE_NAME = "org.jetbrains.kotlinx.lincheck.runner."

--- a/src/jvm/main/org/jetbrains/kotlinx/lincheck/execution/ExecutionResult.kt
+++ b/src/jvm/main/org/jetbrains/kotlinx/lincheck/execution/ExecutionResult.kt
@@ -145,8 +145,8 @@ val ExecutionResult.parallelResults: List<List<Result?>> get() =
 val ExecutionResult.threadsResults: List<List<Result?>> get() =
     threadsResultsWithClock.map { it.map { r -> r.result } }
 
-val ExecutionResult.allResults: Sequence<Result?> get() =
-    initResults.asSequence() + parallelResults.asSequence().flatten() + postResults.asSequence()
+val ExecutionResult.allResults: List<Result?> get() =
+    initResults + parallelResults.flatten() + postResults
 
 // for tests
 fun ExecutionResult.equalsIgnoringClocks(other: ExecutionResult) =

--- a/src/jvm/main/org/jetbrains/kotlinx/lincheck/execution/ExecutionResult.kt
+++ b/src/jvm/main/org/jetbrains/kotlinx/lincheck/execution/ExecutionResult.kt
@@ -145,6 +145,9 @@ val ExecutionResult.parallelResults: List<List<Result?>> get() =
 val ExecutionResult.threadsResults: List<List<Result?>> get() =
     threadsResultsWithClock.map { it.map { r -> r.result } }
 
+val ExecutionResult.allResults: Sequence<Result?> get() =
+    initResults.asSequence() + parallelResults.asSequence().flatten() + postResults.asSequence()
+
 // for tests
 fun ExecutionResult.equalsIgnoringClocks(other: ExecutionResult) =
     this.withEmptyClocks == other.withEmptyClocks

--- a/src/jvm/test/org/jetbrains/kotlinx/lincheck_test/AbstractLincheckTest.kt
+++ b/src/jvm/test/org/jetbrains/kotlinx/lincheck_test/AbstractLincheckTest.kt
@@ -30,6 +30,7 @@ abstract class AbstractLincheckTest(
                 "This test should fail, but no error has been occurred (see the logs for details)"
             }
         } else {
+            checkFailureIsNotLincheckInternalBug(failure)
             failure.trace?.let { checkTraceHasNoLincheckEvents(it.toString()) }
             assert(expectedFailures.contains(failure::class)) {
                 "This test has failed with an unexpected error: \n $failure"

--- a/src/jvm/test/org/jetbrains/kotlinx/lincheck_test/ExceptionTest.kt
+++ b/src/jvm/test/org/jetbrains/kotlinx/lincheck_test/ExceptionTest.kt
@@ -104,6 +104,61 @@ class ExceptionInPostPartTest : AbstractLincheckTest(IncorrectResultsFailure::cl
 
 }
 
+/*
+ * This test checks that Lincheck correctly handles exceptions with empty stack traces.
+ *
+ * JVM can throw exceptions with empty stack traces in some cases as an optimization.
+ * If the same exception (from a predefined list of classes, like `ClassCastException`)
+ * is thrown multiple times from the same line of code,
+ * at some point the JVM will stop collecting the stack trace of this exception
+ * and will return an empty stack trace.
+ *
+ * See:
+ * - https://stackoverflow.com/questions/2411487/nullpointerexception-in-java-with-no-stacktrace
+ * - https://yoshihisaonoue.wordpress.com/2021/02/07/jvm-option-xx-omitstacktraceinfastthrow/
+ * - https://github.com/JetBrains/lincheck/issues/381
+ *
+ * To trigger this JVM behavior, in this test we use a scenario with multiple actors
+ * all throwing `ClassCastException`, and enable the scenario minimization
+ * to force the Lincheck to execute the code multiple times.
+ * At some large enough minimization iteration, the exception will be thrown with an empty stack trace.
+ */
+class EmptyStackTraceExceptionTest : AbstractLincheckTest(IncorrectResultsFailure::class) {
+
+    @Operation
+    fun operation() {
+        var counter = 0
+        while (true) {
+            var obj: Any?
+            if (counter % 2 == 0) {
+                obj = A()
+            } else {
+                obj = B()
+            }
+            // Cause `ClassCastException`
+            obj as A
+
+            counter++
+        }
+    }
+
+    class A
+    class B
+
+    override fun <O : Options<O, *>> O.customize() {
+        iterations(1)
+        threads(1)
+        actorsBefore(0)
+        actorsAfter(0)
+        actorsPerThread(200)
+        minimizeFailedScenario(true)
+        sequentialSpecification(ExceptionTestSequentialImplementation::class.java)
+        logLevel(LoggingLevel.INFO)
+    }
+
+}
+
+
 class ExceptionTestSequentialImplementation {
     fun operation() {}
     fun idle() {}

--- a/src/jvm/test/org/jetbrains/kotlinx/lincheck_test/util/TestUtils.kt
+++ b/src/jvm/test/org/jetbrains/kotlinx/lincheck_test/util/TestUtils.kt
@@ -89,6 +89,10 @@ fun checkTraceHasNoLincheckEvents(trace: String) {
     check(testPackageOccurrences == lincheckPackageOccurrences) { "Internal Lincheck events were found in the trace" }
 }
 
+fun checkFailureIsNotLincheckInternalBug(failure: LincheckFailure) {
+    check("You've caught a bug in Lincheck." !in failure.toString()) { "Internal Lincheck bug was detected" }
+}
+
 /**
  * Indicates whether the current Java Development Kit (JDK) version is JDK 8.
  *

--- a/src/jvm/test/org/jetbrains/kotlinx/lincheck_test/util/TestUtils.kt
+++ b/src/jvm/test/org/jetbrains/kotlinx/lincheck_test/util/TestUtils.kt
@@ -66,11 +66,11 @@ private val String.filtered: String get() {
 
 // removing the following lines from the trace (because they may vary):
 // - pattern `org.jetbrains.kotlinx.lincheck.runner.TestThreadExecution(\d+)` (the number of thread may vary)
-// - everything from `java.base/` (because code locations may vary between different versions of JVM)
+// - everything from `java.base/java.lang` (because code locations may vary between different versions of JVM)
 private val TEST_EXECUTION_TRACE_ELEMENT_REGEX = listOf(
     "(\\W*)at org\\.jetbrains\\.kotlinx\\.lincheck\\.runner\\.TestThreadExecution(\\d+)\\.run\\(Unknown Source\\)",
     "(\\W*)at org\\.jetbrains\\.kotlinx\\.lincheck\\.runner\\.FixedActiveThreadsExecutor\\.testThreadRunnable\\\$lambda\\\$(\\d+)\\(FixedActiveThreadsExecutor.kt:(\\d+)\\)",
-    "(\\W*)at java.base\\/(.*)"
+    "(\\W*)at (java.base\\/)?java.lang(.*)"
 ).joinToString(separator = ")|(", prefix = "(", postfix = ")").toRegex()
 
 private val LINE_NUMBER_REGEX = Regex(":(\\d+)")

